### PR TITLE
Set startup time with env NATS_STARTUP_DELAY

### LINKED ISSUE #3742: Allow to set max startup time on windows service
#### Use Case:

The windows service reports a failure to start if it's not ready to accept connections within 10 seconds.

This value was fixed and hardcoded since before JetStream. On (some) Windows systems, this leads to service startup failures, as the store dir sorting may be hindered by an important load, or slowed down by increased accesses times, typically from security software influence.

see [the relevant line of code](https://github.com/nats-io/nats-server/blob/d44362f696eb942b747da46ce436b60855e23fd9/server/service_windows.go#L70)

#### Proposed Change:

Check for an env var to allow setting this delay to any delay depending on expectations for the current use case. 

#### Who Benefits From The Change(s)?

windows users, esp. when jetstream and security software are competing for processing resources. 

#### Alternative Approaches

Currently setting a NATS service as delayed start, limits occurrences of failures to start. 
Once the streams and consumers are big enough to cause startup failures, several manual starts usually restores a startup time under 10 seconds. 

Another way to work around the service start failure is to whitelist the server storage dir with security software. IT services are usually not happy with this.

### PR DESCRIPTION
**Pull request by Alberic-Hardis**
The startup delay when starting as a service under windows was set to 10 seconds.

This value was fixed and hardcoded since before JetStream. On (some) Windows systems, this leads to service startup failures, as the store dir sorting may be hindered by an important load, or slowed down by increased accesses times, typically from security software influence.

This commit allows overriding the startup delay using an environment variable NATS_STARTUP_DELAY. This variable is checked outside init() to allow reporting into the service log.

Any value can be used (even lower than the initial 10s). Bare NATS start can be far faster than 10s. Ops may want to setup this env var to lower values in order to trigger an error on stricter constraints.

 - [x] Link to issue, e.g. `Resolves #NNN`
 - [ ] Documentation added (if applicable)
 - [ ] Tests added
 - [x] Branch rebased on top of current main (`git pull --rebase origin main`)
 - [ ] Changes squashed to a single commit (described [here](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))
 - [ ] Build is green in Travis CI
 - [x] You have certified that the contribution is your original work and that you license the work to the project under the [Apache 2 license](https://github.com/nats-io/nats-server/blob/main/LICENSE)

Resolves #3742 

### Changes proposed in this pull request:

 - make max startup delay of server a variable
 - set this duration from NATS_STARTUP_DELAY environment variable ("30s" for instance)
 - in case of failure to parse or failure to retrieve value, default to 10 seconds
 - log parsing errors of variable into system log with the winservice instance (for alerting) 

/cc @nats-io/core

Note : I could not add tests for this change, setting the value of the env var  to low delays shows the change appears to have the expected behavior.

### CODE REVIEW

**Comment by Alberic-Hardis** [comment]
Travis build fails, but it cannot be caused by the change. The changes all happen in a windows-only file, and all tests are run under linux focal.

**Review by ColinSullivan1 — APPROVED** [review]
LGTM.

**Review by ColinSullivan1 — COMMENTED** [review]
Thanks!  Looks good to me.  Could you add a test?  Perhaps check the time on startup it would take to accept a connection.

File: server/service_windows.go [hunk 1]
```diff
@@ -66,8 +67,16 @@ func (w *winServiceWrapper) Execute(args []string, changes <-chan svc.ChangeRequ
 	status <- svc.Status{State: svc.StartPending}
 	go w.server.Start()
 
+	if v, exists := os.LookupEnv("NATS_STARTUP_DELAY"); exists {
+		delay, err := time.ParseDuration(v)
```
Comment (derekcollison) [inline server/service_windows.go:71]: `if delay, err := time.ParseDuration(v); err == nil {`

Comment (Alberic-Hardis) [reply server/service_windows.go:71]: yes, I don't see why I did not do it.

**Comment by Alberic-Hardis** [comment]
> Thanks! Looks good to me. Could you add a test? Perhaps check the time on startup it would take to accept a connection.

@ColinSullivan1 I don't know if it's actually a good idea 

- the travis runner does not run tests on windows (this would be useful only for coders on windows)
- this is the implementation of service wrapping for windows, only used when nats-server is installed as a service

I think we could make `winServiceOption func(winServiceWrapper)`, that would both make it future ready (some other behavior might become configurable later on), and allow checking the option application onto the `winService` struct.

Would that be  OK for you ?

**Comment by Jarema** [comment]
It would be good to update the docs too.

**Comment by Alberic-Hardis** [comment]
> It would be good to update the docs too.
@Jarema here you are
https://github.com/nats-io/nats.docs/pull/550

**Comment by Alberic-Hardis** [comment]
Update : 

I tried to write some tests for this, but either I fail to reach the `winServiceWrapper.Execute()` or I just end up writing a test that tests `os.LookupEnv()`, basically. 

Other option is to mock the server to test `winserviceWrapper.Execute`, but that implies redefining `winServiceWrapper` with an interface instead of simply a `*Server`, to make the server mockable. This implies too many changes too unrelated with the issue of #3742. 

I'll give it another try if I have a better inspiration. This definitely makes me realize why no tests were written for this piece of code previously.

**Comment by Alberic-Hardis** [comment]
Happy year 2023 everyone. 
I've copied the logic of the tests in service_test.go to write a windows-specific test (builds only on windows) 

Travis still fails for some unrelated reason (no code in the proposed changes can run in Travis test suite that runs on Linux) 

The tests I propose here use a crude but sufficient windows services mock, and target the delay env var by setting it at the beginning of the test in order to make the startup fail (`winServiceWrapper.Execute` returns `false`) in the short timeframe of the tests.

**Comment by ColinSullivan1** [comment]
@Alberic-Hardis Thanks!  We've run the test on windows (thanks @scottf).  LGTM!

**Comment by Alberic-Hardis** [comment]
Hello all, 
Is this merge likely to make its way into main and an official release ?

**Comment by derekcollison** [comment]
Will be part of 2.10.

**Pull request merged**